Added rawfile_json returner, writes directly to local file

fdamstra opened the pull request:
### What does this PR do?

Adds a simple returner that dumps the json for returned events to a file on the local filesystem for easy parsing by other software. This returner writes to a file on the local filesystem of either the minion (in the event of --return) or of the master (when using event_return).
### What issues does this PR fix or reference?

None
### New Behavior

Filename can be specified by custom variable. If configured as a returner, will output event data in JSON format to a raw file.
### Tests written?

No

fdamstra committed:
Added rawfile_json returner

This returner saves the event json directly to a file on the local
filesystem of either the minion (in the event of --return) or
of the master (when using event_return).

fdamstra committed:
Added documentation link for rawfile_json returner.

cachedout commented on salt/returners/rawfile_json.py:
```diff
@@ -0,0 +1,76 @@
+
+def _get_options(ret):
+    '''
+    Returns options used for the carbon returner.
```
Did this get copied/pasted? ;]

fdamstra replied:
Good eye. I'm new to Python. Fixed in latest set of patches. I appreciate the feedback.

cachedout commented on salt/returners/rawfile_json.py:
```diff
@@ -0,0 +1,76 @@
+    '''
+    opts = _get_options({})  # Pass in empty ret, since this is a list of events
+    try:
+        with open(opts['filename'], 'a') as logfile:
```
Bear in mind that you have multiple processes writing to this file at once and it doesn't look like there's any locking happening here. This might result in processes writing over each other and invalid JSON as a result..

fdamstra replied:
I switched to salt.utils.flopen() which addresses this issue and the next. I wondered if this needed to be threadsafe. Thanks for the feedback.

cachedout commented on salt/returners/rawfile_json.py:
```diff
@@ -0,0 +1,76 @@
+    '''
+    opts = _get_options({})  # Pass in empty ret, since this is a list of events
+    try:
+        with open(opts['filename'], 'a') as logfile:
```
To be certain that the file is closed correctly so that we don't leak file handles, we prefer the use of `salt.utils.fopen`

fdamstra replied:
Thanks. Switched to salt.utils.flopen to fix this and the locking concern.

fdamstra committed:
Switched from open() to salt.utils.flopen()

flopen() provides locking and file handle tracking.
Also fixed errant copy and pasted comment.

fdamstra committed:
Switched from open() to salt.utils.flopen()

flopen() provides locking and file handle tracking.
Also fixed errant copy and pasted comment.

fdamstra commented:
@cachedout: Appreciate all the feedback. 

Something went wrong with my rebase and I've pulled in a number of other peoples' commits. I'm looking into this.

fdamstra committed:
Rebased to develop

fdamstra commented:
There fixed. The rebase so only my changes are reflected in the pull request. Any other comments or concerns?

cachedout commented:
Thanks, @fdamstra !
